src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -10,14 +10,36 @@ namespace Microsoft.ML.Runtime.Internal.CpuMath
     public static partial class CpuMathUtils
     {
         // The count of bytes in Vector128<T>, corresponding to _cbAlign in AlignedArray
-        public const int Vector128Alignment = 16;
+        private const int Vector128Alignment = 16;
+
+        // The count of bytes in Vector256<T>, corresponding to _cbAlign in AlignedArray
+        private const int Vector256Alignment = 32;
+
+        public static int GetVectorAlignment()
+        {
+            // Assumes SSE support on machines that run ML.NET.
+            return Avx.IsSupported ? Vector256Alignment : Vector128Alignment;
+        }
 
         public static void MatTimesSrc(bool tran, bool add, AlignedArray mat, AlignedArray src, AlignedArray dst, int crun)
         {
             Contracts.Assert(mat.Size == dst.Size * src.Size);
             Contracts.Assert(crun >= 0);
 
-            if (Sse.IsSupported)
+            if (Avx.IsSupported)
```

Comment:
We should ensure that using 256-bit operations is a net-win for the entire application before "enabling" it by default. For certain workloads, it can downclock the processor and while this generally results in a higher throughput for the specific 256-bit workload, it can potentially cause a perf hit for any non 256-bit workloads that get executed before the processor goes back to its "highest" frequency level.

As per 15.26 Skylake Server Power Management (from the "Intel® 64 and IA-32 Architectures Optimization Reference Manual")
![image](https://user-images.githubusercontent.com/10487869/44477250-eeae4d00-a5ee-11e8-9282-a3753b041646.png)

Reply:
> It is not always easy to predict whether a program's performance will improve from building it to target
Intel AVX-512 instructions. Programs that enjoy high performance gains from the use of xmm or ymm
registers may expect performance improvement by moving to the use of zmm registers. However, some
programs that use zmm registers may not gain as much, or may even lose performance. It is recommended
to try multiple build options and measure the performance of the program.

The same logic applies, but to a lesser extent, to AVX-256 instructions.

Reply:
@briancylui - there is one end-to-end benchmark in the repo today - https://github.com/dotnet/machinelearning/tree/master/test/Microsoft.ML.Benchmarks.  Can you try running it with/without your changes and post the results?

Reply:
In addition to @tannergooding 's concern, what will be the end-to-end scenarios or applications to drive the optimizations? https://github.com/dotnet/machinelearning/tree/master/test/Microsoft.ML.Benchmarks lists one end-to-end benchmark, but it seems iris.txt is a small data set. Do we have more end-to-end scenarios to test?

Reply:
Yes, we are currently formulating more benchmarks to add to the repo.  See #711.

Reply:
@eerhardt: Feel free to check out the perf results for end-to-end perf scenarios on https://github.com/briancylui/machinelearning/issues/7

Reply:
@tannergooding suggested earlier about checking if the input array length is greater than a certain `AvxLengthLimit` along with checking `Avx.IsSupported`.  I could do some perf tests that accept a `Param` that is the length of the input array to see what value `AvxLengthLimit` should take.  Same for `SseLengthLimit`.

I think one of the problems about AVX that impacts perf right now is that we haven't implemented the "**double-compute**" to take care of alignment that @tannergooding mentioned in https://github.com/briancylui/machinelearning/issues/2 (bottom item).  I look forward to implementing it after this PR, since it would change the layout of every function in a significant way.